src/node/blockstorage.cpp

```diff
@@ -959,8 +960,12 @@ bool BlockManager::WriteBlockUndo(const CBlockUndo& blockundo, BlockValidationSt
                 // Write undo data & checksum
                 fileout << blockundo << hasher.GetHash();
             }
+        }
 
-            fileout.flush(); // Make sure `AutoFile`/`BufferedWriter` go out of scope before we call `FlushUndoFile`
+        // Make sure that the file is closed before we call `FlushUndoFile`.
+        if (file.fclose() != 0) {
+            LogError("Failed to close block undo file %s: %s", pos.ToString(), SysErrorString(errno));
+            return FatalError(m_opts.notifications, state, _("Failed to close block undo file."));
```

Comment:
do we need both messages? How will this be displayed in the logs, isn't it redundant?

Reply:
This just mimics the error handling when the file cannot be opened:

https://github.com/bitcoin/bitcoin/blob/53ccb75f0c47e1e48504cb392d7589548a8d25a2/src/node/blockstorage.cpp#L945-L949